.pre-commit-config.yaml

```diff
@@ -0,0 +1,17 @@
+repos:
+  - repo: https://gitlab.com/pycqa/flake8
```

Comment:
Do you know if flake8 is going to be run on the diff, or on entire folders? I'm asking because flake8 can take a reaaaaally long time to check stuff, especially on MacOS were multiprocessing isn't supported (at least it wasn't until recently)

Ideally we should try to keep the entire set of hooks to run in << 1sec for a reasonable dev experience


Also I'm wondering whether this may actually get in our way, rather than help: some devs have workflows where they commit or amend often with very messy code that would not pass the checks, and only publish when they have a decent version. Personally I never care about flake8 up until I set a PR as ready to review, but until then I don't worry about it.
Considering that most of flake8 issues will be solved by black (which doesn't require any human intervention, unlike flake8), I think it might make sense to remove this one. WDYT?

Reply:
> Do you know if flake8 is going to be run on the diff, or on entire folders?

By default, it will run on any file that was changed or added in the commit. Each hook knows which files it can handle so if you just add python files, the checks for yaml and toml will be skipped. 

> Also I'm wondering whether this may actually get in our way, rather than help: some devs have workflows where they commit or amend often with very messy code that would not pass the checks, and only publish when they have a decent version.

The obvious way out is to simply not use `pre-commit` hooks if you have a different workflow. You can still `pre-commit run` manually to check the current diff. So, IMO we should keep it. Otherwise, contributors need to run multiple commands to get all formatting checks, which is worse.

Reply:
> Most of them yes, but a usual offender is unused imports.

Hm, that's typically the kind of things that can happen reasonably often when iterating and committing often, so it would be best to avoid it IMO.

> The obvious way out is to simply not use pre-commit hooks if you have a different workflow. You can still pre-commit run manually to check the current diff

Running `pre-commit run` woudl still check flake8, right?

> Otherwise, contributors need to run multiple commands to get all formatting checks, which is worse.

I agree it's not ideal, but I don't think anyone runs the flake8 checks locally  at the moment anyway. Our current instructions are to run `flake8 --config=setup.cfg .`: not only does it take about 20sec to finish, but it also will flag the entire repo instead of just checking the diff. I would be surprised if anyone used that. So not having the flake8 checks as part of the hook wouldn't be a regression on that side I think.

Reply:
> Running pre-commit run woudl still check flake8, right?

Yes, but why would you run the checks in the first place if your state is messy and you don't want to fix it? My point is, if you follow the workflow of committing often and messy versions, simply `pre-commit run` when you are done and want to publish the commit(s). 

> I agree it's not ideal, but I don't think anyone runs the flake8 checks locally at the moment anyway. 

I do unless I forget it. This is one of the reasons I want to have it in the hooks. It also saves quite some CI time since this can easily be checked locally.

Reply:
> I do unless I forget it

What command do you run? On my machine, with the recommended command I get

```
(pt) ➜  vision git:(main) ✗ time flake8 --config=setup.cfg . | wc -l
223
flake8 --config=setup.cfg .  115.72s user 23.81s system 755% cpu 18.474 total
wc -l  0.00s user 0.00s system 0% cpu 18.473 total
```

So running flake8 is takes 20seconds and gives me 223 lines of noise. Replacing `.` with `torchvision` things get slightly better, but still unacceptable for a regular workflow

```
(pt) ➜  vision git:(main) ✗ time flake8 --config=setup.cfg torchvision | wc -l
35
flake8 --config=setup.cfg torchvision  58.93s user 12.64s system 744% cpu 9.619 total
wc -l  0.00s user 0.00s system 0% cpu 9.618 total
```

If the pre-commit is on the diff, the time should hopefully be acceptable, but maybe not a big PRs. But it doesn't change the workflow issue:

> Yes, but why would you run the checks in the first place if your state is messy and you don't want to fix it?

Mainly because I would still like to enjoy a well black-formatted code (with the pre-commit hooks) without having to worry about flake8 until I really need to :)

Reply:
> What command do you run?

```bash
$time flake8 --config=setup.cfg .

real    0m1.047s
user    0m11.580s
sys     0m0.327s
```

> So running flake8 is takes 20seconds and gives me 223 lines of noise. Replacing . with torchvision things get slightly better, but still unacceptable for a regular workflow

After some offline discussion, we realized this is a problem with MacOS and Python >=3.8. Still, I agree, 20 seconds is unacceptable. But this is the time for checking everything. A "normal" will hardly touch more than 10 files at the same time so the number should be significantly lower. Could you re-run the checks on your diff the next time you want to publish a commit?

> Mainly because I would still like to enjoy a well black-formatted code (with the pre-commit hooks) without having to worry about flake8 until I really need to

Well, this is a different issue. You will still be able to run `ufmt format .` without touching `flake8`. Even better, `black` has [instructions for a lot of editors and IDEs](https://black.readthedocs.io/en/stable/integrations/editors.html) to run after every save. No need to wait for a commit. I use this setup, so during a commit I only need to run the checks.

Reply:
Checking `torchvision/transforms` which should be an upper bound for most PRs takes a bit more than 1 sec. It's not great but not terrible either. Now, I observed huge variations (up to 20X) depending on the Python version that I was on, so I don't think we should use my setup as a baseline anyway: for some contributors, the flake8 time can be much better and probably can be worse too.

> Well, this is a different issue. You will still be able to run ufmt format . without touching flake8

This is not a different issue IMO, this is *the* issue. I shouldn't have to manually run ufmt, or to perform IDE-specific instructions. That's the point of having commit hooks in the first place :)

Reply:
We probably need to agree to disagree here. For me these hooks are specifically for checking a commit adheres to our standards before its done. Relying on the hooks just to invoke the code formatter seems odd. If you don't care about the style (code format and other things) for intermediate commits, don't use hooks. If you just care about code format for intermediate work, why commit at all at not just `pre-commit run`?

This is not a hill I want to die on, so I'll leave the choice up to you. 

---

As an afterthought: Is this specifically a problem with `flake8` or all hooks that are not self-correcting?

Reply:
> Is this specifically a problem with flake8 or all hooks that are not self-correcting?

Pretty much, yeah. Or at least those that would often not pass in a quick and dirty iterative workflow. So yes, mypy would be one of them :p  

But you make some good points and I would agree that if a repo has pre-commit hooks set up, as contributors we wouldn't expect the CI to go red for a linting issue / something that can be easily caught by hooks.

I also just realized that specific hooks can be consistently skipped with a` SKIP` env variable (https://pre-commit.com/#temporarily-disabling-hooks).

Let's try with flake8 in then, I guess I can just edit my `.zshrc` to always skip it

Thanks for the discussion @pmeier

Reply:
> So yes, mypy would be one of them :p

`mypy` is a entirely different beast, since hook runs in an isolated environment. `mypy` requires the package under inspection as well as all dependencies to be installed, so we would need to tell `pre-commit` to do that. I've tried that with a project of mine that depends on PyTorch and was more pain than to just exclude it and invoke it manually.